### PR TITLE
added formatting when source snippets is too long

### PR DESCRIPTION
**Pull request by adamsau**
reference #2885

the pointers are always display as ^--------^ for simplicity.

### CODE REVIEW

**Comment by pirapira** [comment]
Restarted the emscripten build.

File: libsolidity/interface/SourceReferenceFormatter.cpp [hunk 1]
```diff
@@ -49,6 +49,21 @@ void SourceReferenceFormatter::printSourceLocation(
 	if (startLine == endLine)
 	{
 		string line = scanner.lineAtPosition(_location->start);
+
+		int locationLength = endColumn - startColumn;
+		if (locationLength > 150)
+		{
+			line = line.substr(0, startColumn) + line.substr(startColumn, 15) + "..." + line.substr(endColumn - 15, 15) + line.substr(endColumn, line.length() - endColumn);
```
Comment (chriseth) [inline libsolidity/interface/SourceReferenceFormatter.cpp]: Why don't you use `line = line.substr(0, startColumn + 15) + "..." + line.substr(endColumn - 15);`?

Comment (chriseth) [reply libsolidity/interface/SourceReferenceFormatter.cpp]: Oh and perhaps `" ... "` instead of `"..."` would be better readable.

File: libsolidity/interface/SourceReferenceFormatter.cpp [hunk 2]
```diff
@@ -49,6 +49,21 @@ void SourceReferenceFormatter::printSourceLocation(
...
+			endColumn = startColumn + 33;
+			locationLength = 33;
```
Comment (chriseth) [inline libsolidity/interface/SourceReferenceFormatter.cpp]: Hm, having seen that, perhaps the restriction to 15 is a bit too small, since it means we drop from 149 to 33. Perhaps we should rather cut it in the middle so that the length is never larger than 150 characters?

File: libsolidity/interface/SourceReferenceFormatter.cpp (hunk 1 above)
Comment (adamsau) [reply libsolidity/interface/SourceReferenceFormatter.cpp]: yes you are right :)

File: libsolidity/interface/SourceReferenceFormatter.cpp (hunk 2 above)
Comment (adamsau) [reply libsolidity/interface/SourceReferenceFormatter.cpp]: if it was cut in the middle and take the first half i guess, there is still chance that the half is > 150 chars. If i get the first 150 chars in the half, the total length of the line is likely to be > 150 chars, which means the text around the location is likely to be convert to "... "

Comment (adamsau) [reply libsolidity/interface/SourceReferenceFormatter.cpp]: i suggest if the source location is > 150 chars, we only take the first 75 chars and append " ... " as a shortened source location.

Comment (adamsau) [reply libsolidity/interface/SourceReferenceFormatter.cpp]: @chriseth please check the latest PR.

File: libsolidity/interface/SourceReferenceFormatter.cpp [hunk 3]
```diff
@@ -49,6 +49,21 @@ void SourceReferenceFormatter::printSourceLocation(
 	if (startLine == endLine)
 	{
 		string line = scanner.lineAtPosition(_location->start);
+
+		int locationLength = endColumn - startColumn;
+		if (locationLength > 150)
+		{
+			line = line.substr(0, startColumn + 75) + " ... " + line.substr(endColumn);
```
Comment (chriseth) [inline libsolidity/interface/SourceReferenceFormatter.cpp]: Why did you remove the context before `endColumn`?

Comment (adamsau) [reply libsolidity/interface/SourceReferenceFormatter.cpp]: As mentioned in the previous commit comment , i suggest if the source location is > 150 chars, we only take the first 75 chars and append " ... " as a shortened source location.

Comment (chriseth) [reply libsolidity/interface/SourceReferenceFormatter.cpp]: Sure, but sometimes the end might be more important than the beginning.

Comment (adamsau) [reply libsolidity/interface/SourceReferenceFormatter.cpp]: So any suggestions on how to shorten the source location?

Comment (chriseth) [reply libsolidity/interface/SourceReferenceFormatter.cpp]: keep x characters after start and x characters before end so that the whole `line` string has a length of 150.

Comment (chriseth) [reply libsolidity/interface/SourceReferenceFormatter.cpp]: ah no, sorry. Something like
`line = line.substr(0, startColumn + 70) + " ... " + line.substr(endColumn - 70);`

Comment (chriseth) [reply libsolidity/interface/SourceReferenceFormatter.cpp]: Sorry about the back-and-forth here.

Comment (adamsau) [reply libsolidity/interface/SourceReferenceFormatter.cpp]: I understand that, but if I take the first 70 and last 70 chars of source location and the dots, it will be 145 chars for source location. If there is code around the source location, most likely the total length of line will be > 150, and the code around the source location will be converted to " ... ". Is that fine?

Comment (chriseth) [reply libsolidity/interface/SourceReferenceFormatter.cpp]: Yeah, I see. So what we probably want is some chars left and right of the beginning and end of the source location. So what about 
`line.substr(0, startColumn + 35) + " ... " + line.substr(endColumn - 35)` ?

Comment (adamsau) [reply libsolidity/interface/SourceReferenceFormatter.cpp]: yes that will be good

**Review by chriseth — APPROVED** [review]

**Comment by axic** [comment]
Shouldn't this have a changelog entry? Also can we squash the commits?

**Comment by chriseth** [comment]
Yes, you are right about the changelog entry. Squashing should also be done, but can also be done as part of the merge in github.

**Comment by pirapira** [comment]
@rivenhk will you add a changelog entry?

**Comment by adamsau** [comment]
@pirapira done~

**Comment by pirapira** [comment]
@rivenhk thank you. I restarted the failing test (the failure looked unrelated to the PR).

**Comment by axic** [comment]
@pirapira I hoped the first 3 commits would have been squashed given they do a lot of back-and-forth changes (not you can also do that from the merge interface)

**Comment by pirapira** [comment]
@axic I knew, but I didn't have the chance to do that.  It's already merged.

**Comment by axic** [comment]
@pirapira sorry, it was @chriseth doing it.

**Comment by chriseth** [comment]
Oh I'm sorry, forgot about that...
